contracts/verifier.sol

```diff
@@ -0,0 +1,260 @@
+        returns (uint)
+    {
+        require(steps > 2);
+
+        uint sessionId = sessions.length;
+        sessions.push(VerificationSession({
+            id: sessionId,
+            claimant: claimant,
+            challenger: challenger,
+            input: _input,
+            output: _output,
+            lastClaimantMessage: now,
+            lastChallengerMessage: now,
+            lowStep: 0,
+            lowHash: keccak256(_input),
+            medStep: 0,
+            medHash: 0,
+            highStep: steps,
+            highHash: keccak256(_output)
+        }));
+
+        require(isInitiallyValid(sessions[sessionId]));
+
+        NewClaim(sessionId);
+        return sessionId;
+    }
+
+    modifier onlyClaimant(uint sessionId) {
+        require(msg.sender == sessions[sessionId].claimant);
+        _;
+    }
+
+    // @TODO(shrugs) - this allows anyone to challenge an empty claim
+    //  is this what we want?
+    modifier onlyChallenger(uint sessionId) {
+        var session = sessions[sessionId];
+        if (session.challenger == address(0)) {
+            session.challenger = msg.sender;
+        } else {
+            require(msg.sender == session.challenger);
+        }
+        _;
+    }
+
+    function query(uint sessionId, uint step)
+        onlyChallenger(sessionId)
+        public
+    {
+        VerificationSession storage s = sessions[sessionId];
+
+        bool isFirstStep = s.medStep == 0;
+        bool haveMedHash = s.medHash != bytes32(0);
+        assert(isFirstStep || haveMedHash);
+        // ^ invariant if the step has been set but we don't have a hash for it
+
+        if (step == s.lowStep && step + 1 == s.medStep) {
+            // final step of the binary search (lower end)
+            s.highHash = s.medHash;
+            s.highStep = step + 1;
+        } else if (step == s.medStep && step + 1 == s.highStep) {
+            // final step of the binary search (upper end)
+            s.lowHash = s.medHash;
+            s.lowStep = step;
+        } else {
+            // this next step must be in the correct range
+            require(step > s.lowStep && step < s.highStep);
+
+            // if this is NOT the first query, update the steps and assign the correct hash
+            // (if this IS the first query, we just want to initialize medStep and medHash)
+            if (!isFirstStep) {
+                if (step < s.medStep) {
+                    // if we're iterating lower,
+                    //   the new highest is the current middle
+                    s.highStep = s.medStep;
+                    s.highHash = s.medHash;
+                } else if (step > s.medStep) {
+                    // if we're iterating upwards,
+                    //   the new lowest is the current middle
+                    s.lowStep = s.medStep;
+                    s.lowHash = s.medHash;
+                } else {
+                    // and if we're requesting the midStep that we've already requested,
```

Comment:
I think we should add a revert here.

Reply:
Fixed https://github.com/TrueBitFoundation/scrypt-interactive/pull/24/commits/58e62be4c4e7438c76197fc16533282d3a60d747